### DISCUSSION TITLE
Tasmota Modbus Smart Meter Interface ABB B23 how to newbe question

### DISCUSSION THREAD

**Discussion by matzeeg3** [arendst/Tasmota#21802] [Q&A]
Hi everyone,
i am trying to get a Tasmota running with Scripts enabled. I build my own Firmware for this as documented.
After that i try to Connect my Adapter to Tasmota and use the script provided by the docu website.
I used this script:
`>D
>B
->sensor53 r
>M 1
+1,3,m,0,9600,ABB,1,10,01035B00,01035B02,01035B04,01035B14,01035B16,01035B18,01035B1A,r010350080004,r010350000004,r010350040004
1,010304UUuuUUuu@i0:10,Voltage L1-N,V,Voltage_L1,1
1,010304UUuuUUuu@i1:10,Voltage L2-N,V,Voltage_L2,1
1,010304UUuuUUuu@i2:10,Voltage L3-N,V,Voltage_L3,1
1,010304SSssSSss@i3:100,Active power Total,W,Active_power_Total,2
1,010304SSssSSss@i4:100,Active power L1-N,W,Active_power_L1,2
1,010304SSssSSss@i5:100,Active power L2-N,W,Active_power_L2,2
1,010304SSssSSss@i6:100,Active power L3-N,W,Active_power_L3,2
1,010308xxxxxxxxSSssSSss@i7:100,Real energy,kWh,Real_energy,2
1,010308xxxxxxxxUUuuUUuu@i8:100,Real energy consumed,kWh,Real_energy_consumed,2
1,010308xxxxxxxxUUuuUUuu@i9:100,Real energy delivered,kWh,Real_energy_delivered,2
#`
And my RS485 Adapter is this:
![image](https://github.com/user-attachments/assets/cf453fa1-2c78-48bd-b352-de21e41beafc)
or this:
![image](https://github.com/user-attachments/assets/7af603c3-f1df-4aed-b199-69f940c1121f)
but all the time i am unable to recive data from the SmartMeter.
If i connect the SmartMeter via USB RS485 to my computer i recive the data but not via Tasmota.
Now my Questions: How did i have to wire up the RS485 Adapter, witch on is the best for using with Tasmota and how to chose the correct PINs for it? i use a Wemos Mini (ESP8266).
May Someone can help me in this case?
Thanks in Advanced.
Matze

**Comment by HofeBY** [comment]
In your example you use GPIO3 as RxData, but I don´t understand why you don´t activate internal PullUp Resistor of ESP.
Do you use an externel PullUp Resistor ?
GPIO1 is TxData defined in your example.
1000ms as Period to repeat seems to be OK.
.
RS485 Module is connectet A to A and B to B Signal of 2wire RS485 Bus.
Take care to activate 120Ohm Bus Resistors.
GND an Powersupply connected betwen ESP and RS485 Module.
Most RxData ESP to TxData RS485. 
TxData ESP to RxData RS485.
.
here is the manual for SML:
https://tasmota.github.io/docs/Smart-Meter-Interface/

**Comment by matzeeg3** [comment]
no i don´t have an external pull resistor und i don´t know how to activate the internal pullup.
is the rs485 3.3 volt or 5 volt

**Reply by HofeBY** [reply]
changed your example Line with GPIO3 to active Pullup Resistor.
+1,3,m,1,9600,ABB,1,10,01035B00,01035B02,01035B04,01035B14,01035B16,01035B18,01035B1A,r010350080004,r010350000004,r010350040004

**Comment by matzeeg3** [comment]
i can do what i want with or without the internal pullup i see the RS485 Modul Blinking but i don´t get any data.

**Reply by HofeBY** [reply]
As I expected.
use your first RS485 Module !
all at your own risk !!!

You have two wire RS485 ?
study page 7 of pdf
https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.analog.com/MAX481/datasheet&ved=2ahUKEwi-5-v1u72HAxWn8QIHHQffIewQFnoECAgQAQ&usg=AOvVaw2H4eM5gleHzoYqDBOzBWUO

GND to GND
you have to conntect Vcc to ESP first try 3,3V (It´s a MAX485)
!!! but Page7 VCC means 5V. at your own risk !!! 
Take care not to connect ESP 3,3V Pins to 5V of MAX485 without level shifter.

-RE (Pin2) with DE (Pin3) with DI (Pin4) and than to TXD Gpio1 of ESP

RO (Pin1) to RxD Gpio3 of ESP

If no data is sending, MAX485 is in reading Modus. Halfduplex Modus

page 13 show´s Bus Modus with nessesary 120 Ohm resistors

**Comment by matzeeg3** [comment]
can i use the second modul to?
at the moment i have this one connected

**Reply by HofeBY** [reply]
If there is also a MAX485 included, the pdf is still guilty.

**Reply by matzeeg3** [reply]
it says RS485 to TTL and looks like picture 2 in the first page

**Reply by HofeBY** [reply]
It seem´s that Module 2 is the better solution for 5V and needed 3,3V.

I use MAX3485 Module, because it´s made for 3,3V use.
https://www.ebay.de/itm/265809392901?itmmeta=01J3HJT0J2JM4SQVWKQGBDNBT6&hash=item3de3797505:g:BLEAAOSw2ENi6NOU&itmprp=enc%3AAQAJAAAAwJXhOp1JSMXQqwY7CA6FfSBEUxd0djGKoN6%2FfMteLQgb8bY%2BHHPS3tRCiArueI0ZFaN34UkFh20ongYIWDxYXN5aNJAoDEGSyRA%2B4pFr%2FS2x5uvTYBq5Q60KqYm%2BPWGTsQUYjFYbhAinBhiz2fiwwgE%2BvANz5xssQ%2FTwtxJuv%2FCBGivXH7OjLAaNZL7HF5diAg0loWOgdM12tFpuIlNoLvalqmQNBPbLxqqTJVpjdPcaP%2BoTQE%2FSEXzB0LfzWuExww%3D%3D%7Ctkp%3ABk9SR6SJ6LKcZA
https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.analog.com/media/cn/technical-documentation/data-sheets/1079.pdf&ved=2ahUKEwj50e7Z_b6HAxXNxAIHHX5YAVIQFnoECAkQAQ&usg=AOvVaw0rF9DdVGrzPsdz8YUmRl0r

**Reply by matzeeg3** [reply]
i ordert the modul you recomend hope this one will work :)

**Reply by HofeBY** [reply]
https://www.ebay.de/itm/315356609430?itmmeta=01J3HJT0J1BHHB90HMS4VCXC1S&hash=item496cb7fb96:g:vSgAAOSwlX1j8zII&itmprp=enc%3AAQAJAAAA4OiV9b4ZCDw96KZytwTU914cCiFO0rSfc08kecUbNn9oXAiIovmN1IxnMWyvd5KfsDpxUx2GoOpinPp%2B0cC7VAQ7%2BILN3kOej4teDzQhTctLG8sDQsA5TxrQBqprOOBwACp9NkgijCYdipWhSt3cmxOSZqcxuMWm96y10dlitOg41vDtjMQFKKCQYSMXNU9kEVB7f9SRkfHbPZqYyOvLk6DXDAJAV%2B9oC%2FyEq2G8S%2F2dHsq1SBWWWffp6g%2B216QSHVYkq5bUZxIX%2FNmwUYB6RZfTJvmr1VACJqNCzf%2Bmp7FV%7Ctkp%3ABk9SR6KJ6LKcZA
Module2 seem´s to work. read description !

**Reply by matzeeg3** [reply]
that is the one i use at the moment withount success :(

**Reply by HofeBY** [reply]
Has the ABB been set up so that it also sends data?
https://new.abb.com/products/de/2CMA100164R1000/b23-112-100
->downloads

Is it possible to connect another RS485 transmitter?
You can link the existing modules together and then simulate the data signals without connecting the ESP.
Can you measure voltage with a multimeter?

**Reply by matzeeg3** [reply]
i have a multimeter
if i connect my usb - rs485 adapter i get data from the abb so i think it is configured
i also checked the bout rate and it is set to 9600. can i bring the tasmota in a debug mode?

**Reply by HofeBY** [reply]
that´s the point.
If you can read data from ABB with USB to RS485 Adapter, RS485 Bus seems to be OK. Terminator ?
Remember !
there are only two 120 Ohm on Ends of Bus as terminator Resistors allowed !

Module2 VCC at 3,3V !
set Module 2 on Bus and set TxD to GND ! VCC at 3,3V !
Maybe Oszilloskope ?
RxD LED on Module2 is flashing ?
or RxD LED with 1500 Ohm Resistor to GND ?
Do you get data on Module2 RxD ?

**Reply by matzeeg3** [reply]
at the momtent i have only one 120 ohm resitor at the abb a-b

**Reply by HofeBY** [reply]
hurray, flashing RxD on Module2 is wonderful !!! and ABB is configured in burstmodus.
leave TxD on Module 2 on GND !
connect RxD of Module2 to GPIO3 (RxD) of ESP

take a look at Module 2 nearby A-B Connectors.
There is a Resistor (R9) printed with "121" that means "12" plus 1* "0" outwritten 120 (Ohm)
and maybe it´s activated by connecting "R0" with a soldering point.

new script.txt for ESP don´t forget to add > and # at right place

>D

>B
->sensor53 r

>M 1
+1,3,m,1,9600,ABB,1,10
#


than go to Tasmota Console and send command
sensor53 d1
to dump first Meter
(stop with sensor53 d0)

show us the dump

if not readable change script.txt to inverted GPIO3 by adding "i" bevore GPIO3

>D

>B
->sensor53 r

>M 1
+1,i3,m,1,9600,ABB,1,10
#

and repeat dumpmodus

Use Multimeter and messure Voltage on GPIO1 of ESP. 
If in silent modus shold be 0V.
if it is 3,3V we have to invert GPIO1 by adding an "i" before GPIO1 definition.

**Reply by Katlijk** [reply]
Do you use the right adress id off your meter?

**Reply by matzeeg3** [reply]
> hurray, flashing RxD on Module2 is wonderful !!! and ABB is configured in burstmodus. leave TxD on Module 2 on GND ! connect RxD of Module2 to GPIO3 (RxD) of ESP
> 
> take a look at Module 2 nearby A-B Connectors. There is a Resistor (R9) printed with "121" that means "12" plus 1* "0" outwritten 120 (Ohm) and maybe it´s activated by connecting "R0" with a soldering point.
> 
> new script.txt for ESP don´t forget to add > and # at right place
> 
> > D
> 
> > B
> > ->sensor53 r
> 
> > M 1
> > +1,3,m,1,9600,ABB,1,10
> 
> than go to Tasmota Console and send command sensor53 d1 to dump first Meter (stop with sensor53 d0)
> 
> show us the dump
> 
> if not readable change script.txt to inverted GPIO3 by adding "i" bevore GPIO3
> 
> > D
> 
> > B
> > ->sensor53 r
> 
> > M 1
> > +1,i3,m,1,9600,ABB,1,10
> 
> and repeat dumpmodus
> 
> Use Multimeter and messure Voltage on GPIO1 of ESP. If in silent modus shold be 0V. if it is 3,3V we have to invert GPIO1 by adding an "i" before GPIO1 definition.

i recive the following if i try to edit the script like you mentiont:
12:12:28.197 CMD: sensor53 d1
12:12:28.200 CMD: Grp 0, Cmd 'SENSOR', Idx 53, Len 2, Pld -99, Data 'd1'
12:12:28.203 RSL: RESULT = {"Command":"Error"}

my script is:

1>D
1>B
->sensor53 r
1>M 1
+1,3,m,1,9600,ABB,1,10
1#

without 1 in front of the line

**Reply by matzeeg3** [reply]
with the original script i recive:
12:16:45.090 CMD: sensor53 d1
12:16:45.095 CMD: Grp 0, Cmd 'SENSOR', Idx 53, Len 2, Pld -99, Data 'd1'
12:16:45.099 RSL: RESULT = {"Time":"2024-07-24T12:16:45","SML":{"CMD":"dump: 1"}}
12:16:45.255 > 01 03 5b 16 00 02 36 eb 
12:16:46.256 > 01 03 5b 18 00 02 57 28 
12:16:47.252 > 01 03 5b 1a 00 02 f6 e8 
12:16:48.322 > 01 03 50 08 00 04 d4 cb 
12:16:49.326 > 01 03 50 00 00 04 55 09 
12:16:50.238 RSL: STATE = {"Time":"2024-07-2:16:50.245 RSL: SENSOR = {"Time":"2024-07-24T12:16:50","ABB":{"Voltage_L1":0.0,"Voltage_L2":0.0,"Voltage_L3":0.0,"Active_power_Total":0.00,"Active_power_L1":0.00,"Active_power_L2":0.00,"Active_power_L3":0.00,"Real_energy":0.00,"Real_energy_consumed":0.00,"Real_energy_delivered":0.00}}
12:16:50.361 > 01 03 50 04 00 04 14 c8 
12:16:51.353 > 01 03 5b 00 00 02 d7 2f 
12:16:52.355 > 01 03 5b 02 00 02 76 ef 
12:16:53.415 > 01 03 5b 04 00 02 96 ee 
12:16:54.421 > 01 03 5b 14 00 02 97 2b 
12:16:55.510 > 01 03 5b 16 00 02 36 eb 
12:16:56.453 > 01 03 5b 18 00 02 57 28 
12:16:57.269 CMD: sensor53 d0

**Reply by matzeeg3** [reply]
> Do you use the right adress id off your meter?

Meter set to ID1

**Reply by HofeBY** [reply]
You have to reboot ESP every time you changed something in configuration !
TxD of Module2 is still on GND ?
What´s the Voltage on GPIO1 of ESP ?
There is a reason why I don´t want that nothing is sent out to RS485 Bus !
Are you shure, you are using a Tasmota Version with SML included ?
https://github.com/HofeBY/Tasmota-Firmware
only Versions without RULE in Name.

make tomorrow bevore sunrise a neu dump and show us to compare.
I expect in dump
4Byte Meter Register and than 4Byte Value
